File: leetcode/200. Number of Islands/200._Number_of_Islands_2.java
public class Solution {
    int m, n;
    int[] dx = {-1, 1, 0, 0};
    int[] dy = {0, 0, -1, 1};
    public int numIslands(char[][] grid) {
        m = grid.length;
        n = (m == 0) ? 0 : grid[0].length;
        if( m == 0) return 0;
        int[][] mark = new int[m][n];
        
        int id = 1;
        for(int i=0; i<m; i++) {
            for(int j=0; j<n; j++) {
                if( mark[i][j] == 0 && grid[i][j] == '1')
                    extend(mark, grid, i, j, id++);
            }
        }
        return (id-1);
    }
    public void extend(int[][] mark, char[][] grid, int row, int col, int id) {
        if( row < 0 || row >= m || col < 0 || col >= n || grid[row][col] != '1' || mark[row][col] != 0)
            return;
        mark[row][col] = id;
        for(int i=0; i<4; i++) {
            extend(mark, grid, row+dx[i], col+dy[i], id);
        }
    }
}

File: leetcode/200. Number of Islands/200._Number_of_Islands_2.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        m = len(grid)
        if m == 0:
            return 0
        n = len(grid[0])
        
        ans = 0
        for i in xrange(m):
            for j in xrange(n):
                if grid[i][j] == '1':
                    self.walk(grid, i, j)
                    ans += 1
        return ans
    
    def walk(self, grid, i, j):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == '0':
            return;
        grid[i][j] = '0'    #mark as walked
        directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        for l in directions:
            self.walk(grid, i+l[0], j+l[1])